feat: table aliases

dtbuchholz opened the pull request:
## Summary

Adds the ability to use table aliases in `create`, `write`, and `read` commands, including the `shell`, and also configure a file path with `init`. Closes https://github.com/tablelandnetwork/js-tableland-cli/issues/403.

Edit: also includes support for `controller`, `schema`, and `info`.

## Details

- Adds a global `--aliases`, `-a` option where you specify a JSON file path (checks for validity) or reads from `TBL_ALIASES` / `.tablelandrc.json`.
- `init` provides the ability to, optionally, configure the path to an existing table aliases file or create a new file at a specified directory path—thus, accessible in `.tablelandrc.json`. _Creating_ an aliases file is only possible with `init` and defaults to the name `tableland.aliases.json` in the supplied dir path.
- For example, you can create a table like the following and subsequently use `mytable` in read/write commands:
   ```
   tableland create "CREATE TABLE mytable (id INT);" -a ./tableland.aliases.json
   tableland write "INSERT INTO mytable VALUES (1);" -a ./tableland.aliases.json
   tableland read "SELECT * FROM mytable;" -a ./tableland.aliases.json
   ```
- Same logic applies to the shell and the others mentioned above. No more typing out `prefix_chainId_tableId`! And the shell will log the alias name after each command, along with the existing setup that logs the table name.

## How it was tested

Added tests for `create`, `write`, `read`, `shell`, `controller`, `info`, and `schema` that check for passing a valid aliases file, the updates to it, and also cases where an invalid path is provided. To make sure the logic work, I also added ENS tests where there was potential conflicting logic with aliases.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

joewagner commented on src/lib/commandSetup.ts:
```diff
@@ -129,11 +130,21 @@ export class Connections {
+      if (!isValid) {
+        throw new Error(`invalid table aliases file`);
+      }
+      aliasesNameMap = helpers.jsonFileAliases(aliases);
```
This SDK helper doesn't really work how we want, i.e. it overwrites the whole file.  Also, I think we are going to have to remove the helper from the SDK since it doesn't work in the browser.
As a stop-gap until we figure out the where this helper should go, would you be ok just writing an aliases object that updates the json file instead of overriding it?

dtbuchholz replied:
sure, makes sense!

joewagner commented on src/utils.ts:
```diff
@@ -20,8 +22,8 @@ export const wait = (timeout: number) =>
-    if (chain.includes("goerli")) {
-      return `https://goerli.etherscan.io/tx/${hash}`;
+    if (chain.includes("sepolia")) {
+      return `https://sepolia.etherscan.io/tx/${hash}`;
```
nice!

joewagner reviewed (commented):
@dtbuchholz This is excellent!  This feature makes using the CLI way nicer.
Left a comment on the sdk aliases helper, and also looks like the tests don't pass.
I tried them locally to make suer it's not a CI thing, and I get the same failure locally.

dtbuchholz commented:
> I tried them locally to make suer it's not a CI thing, and I get the same failure locally.

hm, weird. i'll try reworking them. for the sake of GH notifications, should i just move this to draft while i toy with different test setups? (i'm not sure if moving to a draft makes a difference...so maybe that's a moot point.)

joewagner commented:
> @joewagner hm, weird. i'll try reworking them. for the sake of GH notifications, should i just move this to draft while i toy with different test setups? (i'm not sure if moving to a draft makes a difference...so maybe that's a moot point.)

I don't think draft matters the way I've got notifications setup.

joewagner commented:
> @joewagner hm, weird. i'll try reworking them. 

At a glance, it looks like your using a package that creates temp files that look like `/private/var/folders/0k/qb34cwsx0yvf6q6c9cmz26780000gn/T/ee9501348e8f2d55275e44ada49c6a94`.  
That's not a `.json` file, which might be required?

dtbuchholz commented:
> At a glance, it looks like your using a package that creates temp files

Ah, good catch! I could have sworn that was working before but is clearly throwing a specific invalid file error--at least that part of it is expected/working properly!

dtbuchholz commented on src/commands/info.ts:
```diff
@@ -18,32 +20,69 @@ export const builder: CommandBuilder<{}, Options> = (yargs) =>
 
-    const { ens, validator } = await setupCommand({
+    // Note: should `setupCommand` be used again, or instantiate a validator directly?
+    const { validator } = await setupCommand({
```
one question i had re: `setupCommand`—should it only be used once? i.e., i use it above with:
```
const { ens } = await setupCommand({
  ...argv,
});
```
instead, i can instantiate one or the other directly vs using the method.

joewagner replied:
I think this can all be reworked to simply check if an aliases arg was provided with an alias for name, if so do the swap.  Then leave the rest of the logic in this command unchanged.  This would be a convenient spot to use the helper described in the comment on the controller command.

dtbuchholz replied:
i ended up keeping duplicative `setupCommand` calls for when ENS is enabled. the previous logic was using this:
```js
let { name } = argv;
const [tableId, chainId] = name.split("_").reverse();
const parts = name.split("_");
// error if parts are invalid, or if ENS is not enabled
// ...
const { ens, validator } = await setupCommand({
   ...argv,
  chain: parseInt(chainId) as any,
});
```
since an ENS name wont have a `chainId` parsed, using `parseInt(chainId)` won't work properly. thus, i had to use `setupCommand` above to get the table name from ENS, and then get the `chainId` before calling `setupCommand` again for the validator.

dtbuchholz commented on src/lib/commandSetup.ts:
```diff
@@ -127,6 +127,7 @@ export class Connections {
     if (chain) {
       try {
         this._network = helpers.getChainInfo(chain);
+        /* c8 ignore next 3 */
```
not sure why since i didn't touch this, but the coverage wasn't passing due to this...which doesn't look like it was an issue before

joewagner replied:
weird...
There's a coverage report that you can look at in your browser.  If you checkout main and run `npm run coverage:report` you might be able to get more detail.

dtbuchholz commented on src/commands/controller.ts:
```diff
@@ -97,7 +139,21 @@ export const builder: CommandBuilder<{}, Options> = (yargs) =>
-
+          // Check if the passed `name` is valid, otherwise, if it's a table alias
+          try {
+            await globalThis.sqlparser.validateTableName(name);
```
i use this try/catch pattern a few times. i feel like i'm overdoing it and wanted to double check if this is the right approach, or if i should let the original error from `validateTableName` bubble up. (i'm not sure what the general rule of thumb is.)

joewagner replied:
For try/catch in general, my preference is to use it as a means to ensure the user gets an error that they can understand.  If during expected running of the cli this results in an error that is confusing, then you can use a try catch and create an error that says something useful.  Here's an example: https://github.com/tablelandnetwork/js-tableland-cli/blob/6b4923d924c0706f9584dc84bc23062f1df3027a/src/utils.ts#L56-L60

As an aside, Carson did some great work configuring the use of [error cause](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause), which is awesome imo.  We probably haven't fully taken advantage of that yet here. That Mozilla article is worth a read if you haven't already.

For the sqlparser specifically, this kind of usage is already setup to provide a nice error for the user.  So the try catch you added here is concealing our standard invalid table name error and creating a one off.  Which is not ideal imo.
Here's the sqlparser source if you want to look at how that works: https://github.com/tablelandnetwork/go-sqlparser/blob/main/cmd/wasm/main.go#L68
As you can see the error says something like, "error validating name: {details on name}".  This is the same error the Validator will get/give, so it's consistent across the whole product set.

RE:  the aliases feature.
It looks like the logic here is:  
 - If you can't validate the name arg, and there is an alias, don't error
 - proceed as normal with whatever name happens to be in the name variable

In this case you're not always validating the uuTableName, which we weren't doing before, but why validate it sometimes?

My thinking is that in general, to enable aliases in the CLI the logic we want to use is:
 - if there's an alias, replace the user specific tableName with the uuTableName
 - proceed as normal with the uuTableName.

In this case, the internals of the command are always the same, the only difference is that if the user wants to map via the aliases file we enable them to do that.

As an example, I'd replace the changes from line 142 to 156 with:
```
if (aliases) {
  const nameMap = await jsonFileAliases(aliases).read();
  const uuName = nameMap[name];
  name = uuName ?? name;
}

await globalThis.sqlparser.validateTableName(name);
```
Or something like that. 
You could probably create a helper that does that logic, and use it in several other places.  That kind of helper is also nice because it's a pattern that helps you to "never mutate state" (a.k.a. never use `let` in js).

Sorry for the novella length response haha, hopefully it's useful 🚀

dtbuchholz commented:
note: i ended up adding alias support for all of the other methods that use table names

joewagner commented:
@dtbuchholz I took another pass on reviewing all this, and seems like we should avoid changing the command logic and just do the alias mapping if it exists.
See the (really long) comment on the controller command changes.

dtbuchholz commented:
@joewagner that all makes sense, and i really appreciate all of the detail! very helpful...will refactor with those recommendations

dtbuchholz commented:
@joewagner i've (finally) updated this with your recommendations. basically, wherever possible, i reverted the complicated logic & unneeded errors to the original implementation. most of it uses a `getTableNameWithAlias` helper that does what you [noted above](https://github.com/tablelandnetwork/js-tableland-cli/pull/404#discussion_r1276595352). (note: i rebased this to main, which is ahead of staging—so there are some changes from main showing in the diff.)

also...it looks like CI issues with the tests aborting early, again ); (but at least i have some circle CI credits coming soon)

joewagner commented:
@dtbuchholz with a quick glance, this looks good. 
I went ahead and pointed this at main just to see what the diff is.  I'll take a closer look tomorrow, and get the staging branch updated.
The switch to the monorepo is still in limbo, but regardless how that pans out we can get this out on Wednesday.

joewagner commented:
This repo is being depreciated. Closing as this PR has been merged in the new repo
https://github.com/tablelandnetwork/tableland-js/pull/20
